### DISCUSSION TITLE
Does mieru define its own sharing link format?

### DISCUSSION THREAD

**Discussion by PharosVip** [enfein/mieru#60] [Ideas]
I want to support the mieru protocol in my new Android Routen proxy app, and later I will also bring mieru support into Pharos Pro (iOS).
Do you define a format like mieru://xxx or something like that?

**Comment by ayanami-desu** [comment]
据我所知，并没有。

**Comment by enfein** [comment]
We don't have that, but it should be an easy feature to build. We can target 1.13 release before mieru v1 reaching to end of life.

BTW if you package the whole binary in the mobile app, migrating to mieru v2 will be super easy because the configuration will be backward compatible.

**Reply by PharosVip** [reply]
> We don't have that, but it should be an easy feature to build. We can target 1.13 release before mieru v1 reaching to end of life.
> 
> BTW if you package the whole binary in the mobile app, migrating to mieru v2 will be super easy because the configuration will be backward compatible.

Okay, so I'll start with the protocol integration with the current version.

**Reply by enfein** [reply]
> > We don't have that, but it should be an easy feature to build. We can target 1.13 release before mieru v1 reaching to end of life.
> > BTW if you package the whole binary in the mobile app, migrating to mieru v2 will be super easy because the configuration will be backward compatible.
> 
> Okay, so I'll start with the protocol integration with the current version.

1.13 is now available. To get a share link, use `mieru export config` command, which generates a URL like `mieru://xxx`. To import the settings, use `mieru import config <URL>` command.

**Reply by PharosVip** [reply]
Great.

**Comment by PharosVip** [comment]
To make it easier for users to use it, I defined a mieru link with the following format template:

mieru://server:port?username=john&password=secret&network=tcp#GoogleVps

I'll be supporting this format for links in the Routen App, if you have any suggestions or comments let me know.
So far the Routen app has been updated to version 0.0.6, and when it reaches version 0.1.0 I will post a download of the software here.

**Reply by PharosVip** [reply]
I've exposed an interface on top of mieru that facilitates client access:

```
func Mieru(sport, lport int, network, server, user, password string) {
	proxyConfigs := make([]socks5.ProxyConfig, 0)
	var hashedPassword []byte

	hashedPassword = cipher.HashPassword([]byte(password), []byte(user))

	proxyHost := server
	if network == "tcp" {
		proxyConfigs = append(proxyConfigs, socks5.ProxyConfig{
			NetworkType: "tcp",
			Address: util.MaybeDecorateIPv6(proxyHost) + ":" + strconv.Itoa(sport),
			Password: hashedPassword,
			Dial: tcpsession.DialWithOptionsReturnConn,
		})
	} else {
		proxyConfigs = append(proxyConfigs, socks5.ProxyConfig{
			NetworkType: "udp",
			Address: util.MaybeDecorateIPv6(proxyHost) + ":" + strconv.Itoa(sport),
			Password: hashedPassword,
			Dial: udpsession.DialWithOptionsReturnConn,
		})
	}

	udpsession.SetGlobalMTU(9000)

	socks5Config := &socks5.Config{
		UseProxy: true,
		ProxyConf: proxyConfigs, }
	}
	socks5Server, err := socks5.New(socks5Config)
	New(socks5Config) if err ! = nil {
	}

	var socks5Addr string
	socks5Addr = util.MaybeDecorateIPv6(util.LocalIPAddr()) + ":" + strconv.Itoa(lport)

	l, err := net.Listen("tcp", socks5Addr)
	l, err := net.Listen("tcp", socks5Addr) if err ! = nil {
		log.Fatalf("listen on socks5 address tcp %q failed: %v", socks5Addr, err)
	}
	log.Infof("mieru client socks5 server is running")
	if err = socks5Server.Serve(l); err ! = nil {
		log.Fatalf("run socks5 server failed: %v", err)
	}
}

```
So users of mieru only need to provide a few parameters in the link to be imported into the client.

**Reply by PharosVip** [reply]
[Download  Routen 0.0.9 here](https://t.me/Pharos_Pro_Announcements/90)

**Comment by dyhkwong** [comment]
Any chance for an "official" human-readable share link standard? The current protobuf format link is not suitable for others to generate and parse.

**Reply by enfein** [reply]
Added a new feature request: https://github.com/enfein/mieru/issues/176

**Reply by dyhkwong** [reply]
Thanks. In fact what I mean is to propose a standard like `scheme://username:password@host:port?param1=value1&param2=value2` to exchange necessary info between GUI clients, it is not a must to intergrate into mieru. I just don't want to generate a link that can be only recognized by myself and no other software can recognize.

**Reply by yjd** [reply]
> Added a new feature request: #176

mieru export config simple 
Encountering passwords with “.” characters, an error occurs.
see https://en.wikipedia.org/wiki/Percent-encoding

**Reply by enfein** [reply]
I understand the current rule is too strict and doesn't compliant with RFC 3986. This can be relaxed in the future.
